libraries/standard/mqtt/utest/mqtt_state_utest.c

```diff
@@ -0,0 +1,434 @@
+
+    /* Invalid QoS. */
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_SEND, 3 ), MQTTStateNull );
+}
+
+/* ========================================================================== */
+
+void test_MQTT_UpdateStatePublish( void )
+{
+    MQTTContext_t mqttContext = { 0 };
+    const uint16_t PACKET_ID = 1;
+    MQTTStateOperation_t operation = MQTT_SEND;
+    MQTTQoS_t qos = MQTTQoS0;
+    MQTTPublishState_t state;
+    int i = 0;
+
+    /* QoS 0. */
+    state = MQTT_UpdateStatePublish( &mqttContext, 0, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTPublishDone );
+
+    /* Invalid parameters. */
+    /* Invalid ID. */
+    qos = MQTTQoS1;
+    state = MQTT_UpdateStatePublish( &mqttContext, 0, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+    /* NULL context. */
+    state = MQTT_UpdateStatePublish( NULL, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+    /* No record found. */
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+    /* QoS mismatch. */
+    addToRecord( mqttContext.outgoingPublishRecords, 0, PACKET_ID, MQTTQoS2, MQTTPublishSend );
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+
+    /* Invalid transition. */
+    addToRecord( mqttContext.outgoingPublishRecords, 0, PACKET_ID, MQTTQoS1, MQTTPubAckPending );
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+
+    /* Invalid QoS. */
+    operation = MQTT_SEND;
+    addToRecord( mqttContext.outgoingPublishRecords, 0, PACKET_ID, 3, MQTTPublishSend );
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, 3 );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+    operation = MQTT_RECEIVE;
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, 3 );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+
+    /* Invalid current state. */
+    operation = MQTT_SEND;
+    addToRecord( mqttContext.outgoingPublishRecords, 0, PACKET_ID, qos, MQTTStateNull );
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+
+    /* Collision. */
+    operation = MQTT_RECEIVE;
+    addToRecord( mqttContext.incomingPublishRecords, 0, PACKET_ID, MQTTQoS1, MQTTPubAckSend );
+    state = MQTT_UpdateStatePublish( &mqttContext, PACKET_ID, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
+
+    /* No memory. */
+    operation = MQTT_RECEIVE;
+    for( i = 0; i < MQTT_STATE_ARRAY_MAX_COUNT; i++ )
+    {
+        mqttContext.incomingPublishRecords[ i ].packetId = 2;
+        mqttContext.incomingPublishRecords[ i ].qos = MQTTQoS1;
+        mqttContext.incomingPublishRecords[ i ].publishState = MQTTPublishSend;
+    }
```

Comment:
I see this logic getting repeated in line 83. Can this be made a function?

Reply:
Created a new function